### PR TITLE
Add improvement to the casbin auth implementation and grouping policy

### PR DESCRIPTION
**Pull request by zhouzhuojie**
- Refactor `repo.entImpl` so that the files are split into 3 parts
- `pkg/repo/ent_impl_casbin.go` has all the logic related to Auth enforcement
- Added tests for `POST "/api/v1/facts/noscope"`
- Publishable token implementation will be the next PR

### CODE REVIEW

File: pkg/repo/ent_impl.go [hunk 1]
```diff
@@ -0,0 +1,118 @@
+package repo
```
Comment (zhouzhuojie) [inline pkg/repo/ent_impl.go:1]: refactoring

File: pkg/repo/ent_impl_casbin_test.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+		assert.NoError(t, err)
+		err = entImpl.enforcer.LoadPolicy()
+		assert.NoError(t, err)
+
+		shouldPass, err := entImpl.enforcer.Enforce(hashedToken, "example.com", "/api/v1/facts", "POST")
+		assert.True(t, shouldPass)
+		assert.NoError(t, err)
+	})
+
+	t.Run("POST to /api/v1/facts", func(t *testing.T) {
```
Comment (zhouzhuojie) [inline pkg/repo/ent_impl_casbin_test.go:33]: example of adding policy for `"/api/v1/facts"`

File: pkg/repo/ent_impl_casbin_test.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+		shouldPass, err := entImpl.enforcer.Enforce(hashedToken, "example.com", "/api/v1/facts", "POST")
+		assert.True(t, shouldPass)
+		assert.NoError(t, err)
+
+		shouldPass, err = entImpl.enforcer.Enforce(hashedToken, "example.com", "/api/v1/facts/:id", "GET")
+		assert.False(t, shouldPass)
+		assert.NoError(t, err)
+	})
+
+	t.Run("POST to /api/v1/facts/noscope", func(t *testing.T) {
```
Comment (zhouzhuojie) [inline pkg/repo/ent_impl_casbin_test.go:55]: example of adding policy for "/api/v1/facts/noscope"

Comment (zhouzhuojie) [reply pkg/repo/ent_impl_casbin_test.go:55]: @jperichon 

A detailed explanation of what's going on for the authorization framework casbin. This URL is just an example, I'm testing here is the flexibility of enabling "Policy -> Enforcement" of authorization.

The convention we defined here
- https://casbin.org/docs/en/rbac-with-domains
- https://github.com/casbin/casbin/blob/master/examples/rbac_with_domains_model.conf
- OPV's own model is defined here https://github.com/open-privacy/opv/blob/main/pkg/repo/repo_ent_impl.go#L31-L49

Definitions:
```
[request_definition]
r = sub, dom, obj, act

[policy_definition]
p = sub, dom, obj, act, eft
```


The authorization framework
```
sub (i.e, subject, is the **hashed token string**)
dom (i.e, domain, is 1:1 mapped to **OPV's domain**, and facts,scopes,and grant token live in the domain)
obj (i.e., object, correctly we directly use **URL path** as the object for authorization)
act (i.e., action, OPV correctly map HTTP methods as the actions. In policy definition, it's defined as oneof `*`, `GET`, `POST`, `PUT`, and etc. And later merged as a regex string like `(GET)|(POST)`)
eft (i.e. effect, usually it's "allow", and we can also specify "deny" to have denylist style of policies)
```


Therefore, for example, to enable publishable tokens, we need to have a policy for each new grant token created. 

```
p, hashed_token_value_123, example.com, /api/v1/facts/noscope, POST, allow
```

The advance usage can be defined as a group role `g` `publishable_token`, and add tokens to this group.
```
p, publishable_token, example.com, /api/v1/facts/noscope, POST, allow


g, publishable_token, hashed_token_value_123, example.com
g, publishable_token, hashed_token_value_456, example.com
# We can add more tokens to this group publishable_token
```

File: pkg/config/env.go [hunk 4]
```diff
@@ -19,7 +19,7 @@ var ENV = struct {
 	DBDriver             string        `env:"OPV_DB_DRIVER" envDefault:"sqlite3"`
 	DBSetupRetryAttempts uint          `env:"OPV_DB_SETUP_RETRY_ATTEMPTS" envDefault:"9"`
 	DBSetupRetryDelay    time.Duration `env:"OPV_DB_SETUP_RETRY_DELAY" envDefault:"100ms"`
-	DBConnectionStr      string        `env:"OPV_DB_CONNECTION_STR" envDefault:"_opv.sqlite?cache=shared&_fk=1"`
+	DBConnectionStr      string        `env:"OPV_DB_CONNECTION_STR" envDefault:"file:memdb1?mode=memory&cache=shared&_fk=1"`
```
Comment (zhouzhuojie) [inline pkg/config/env.go:22]: changed to default in-memory sqlite for testing. Running local `make run` will use a local https://github.com/open-privacy/opv/blob/1226f254db1a8a7badbac1c8fcbc16b8e7d1c9ab/Makefile#L28

File: pkg/dataplane/dataplane.go [hunk 5]
```diff
@@ -33,7 +33,7 @@ func MustNewDataPlane() *DataPlane {
 	dp.Hasher = crypto.MustNewHasher()
 	dp.Validator = validator.New()
 
-	repo, enforcer, err := repo.NewRepoEnforcer()
+	repo, enforcer, err := repo.NewRepoEnforcer(dp.Logger)
```
Comment (zhouzhuojie) [inline pkg/dataplane/dataplane.go:36]: dependency injected a logger just in case repo may need to log something.

File: pkg/repo/repo.go [hunk 6]
```diff
@@ -20,11 +21,33 @@ const (
...
+
 // Enforcer is an interface that enforces the authz access
 // e.g. Enforce(sub, dom, obj, act)
 type Enforcer interface {
-	AddPolicy(params ...interface{}) (bool, error)
-	Enforce(rvals ...interface{}) (bool, error)
+	AddPolicy(AuthzPolicy) (bool, error)
+	Enforce(AuthzRequest) (bool, error)
+	AddGroupingPolicy(AuthzGroupingPolicy) (bool, error)
 }
```
Comment (zhouzhuojie) [inline pkg/repo/repo.go:51]: refactored the Authz enforcer interface to be more precise about what's the expected input.

**Review by jperichon — APPROVED** [review]

**Pull request merged**